Convert test_flows.sh to python (part 2)

arnout opened the pull request:
More tests converted to python, with additional helpers.

arnout committed:
tests: send_CAPI_command: add tlv class

The dev_send_1905 function needs a list of TLVs as arguments. Up to now,
they were given as tuples, which is a bit hard to understand for the
developer.

Create a tlv class and use that as the arguments to dev_send_1905. It
allows us to move the format_tlv local function to the tlv class.

Update test_flows.py accordingly. It's just a matter of calling the tlv
constructor instead of a simple tuple.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: send_CAPI_command: dev_get_parameter: allow additional parameters

Some instances of dev_get_parameter require additional parameters in the
command, for example to get the bssid corresponding to a specific ssid
on a radio. Add the "additional_parameters" argument to support this.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests_flows: move ap_config_renew test to python

In the shell version, the gateway mac address was converted to hex in
the dev_send_1905 command. That was just there for historical reasons
though, so we no longer do that.

On the other hand, the conversion of a MAC address to uppercase still is
relevant because it was a regression at some point, so keep that.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: move ap_config_bss_tear_down test to python

Since the uppercase MAC address is already tested in ap_config_renew,
there is no need to do it here as well.

To make sure each test stands on its own, use a different SSID than in
ap_config_renew.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: ap_config_bss_tear_down: verify connection map

The original test in test_flows.sh had a kludgy check of bml_conn_map to
verify that the VAPs were properly updated in the controller.

Make an improved version of this check: the get_conn_map() function
parses the bml_conn_map output into a dictionary in which we can look up
nodes. Add the helper classes map_vap, map_radio and map_node. Because
the bml_conn_map output is a bit annoying, we only parse the repeaters
for the time being.

With this improved connection map parsing, verify that everything is
torn down except the configured SSID.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: add helper function to send BWL event

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented on tests/send_CAPI_command.py:
```diff
@@ -15,6 +15,54 @@ class CAPIReply(Enum):
+        self.length = length
+        self.value = value
+
+    def format(self, tlv_num: int = 0) -> str:
```
This format function does something different than simply make a string representation, so NACK. I could add a __str__ that prints the three members though.

arnout commented on tests/test_flows.py:
```diff
@@ -225,6 +295,82 @@ def test_initial_ap_config(self):
+        repeater1 = conn_map[self.mac_repeater1]
+        repeater1_wlan0 = repeater1.radios[self.mac_repeater1_wlan0]
+        for vap in repeater1_wlan0.vaps.values():
+            if vap.ssid != b'N/A':
```
SSIDs can contain any bytes, not limited to UTF-8. So it should most definitely be `bytes`.

arnout commented:
>     * Consider making the tlv class a named tuple (or a subclass of one).

I don't see how a namedtuple is useful here... IMO it is way more awkward than a class.

I would love to use [attrs](https://github.com/hynek/attrs/) but that's unfortunately not a standard Python module so I wouldn't want to rely on it in the top-level test infra.

>     * I don't like `send_CAPI_command.py` as a filename, I'd rather have something like `base_classes.py` (or simply `base.py`). Or maybe call it `capi.py` and put the tlv class in a seperate file

I fully agree, but since send_CAPI_command.py is called from the existing test_flows.sh, I'd prefer to do that rename when test_flows.sh is removed.

arnout committed:
tests: move client_capability_query to python

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: move client_association_dummy to python

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: send_CAPI_command.py: fix flake8 errors

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
I'll continue converting more tests, but it can be merged as is as well so ready for review.

LiorAmram commented on tests/send_CAPI_command.py:
```diff
@@ -16,14 +16,63 @@ class CAPIReply(Enum):
+            The string representation of the TLV according to CAPI definition.
+        """
+        if tlv_num:
+            tlv_num = str(tlv_num)
```
I think it's bad practice to change the type of a variable, especially when you explicitly mark it as int in the method signature. 
I would prefer defining a tlv_num_str variable.

arnout replied:
Good point. Done.

LiorAmram commented on tests/send_CAPI_command.py:
```diff
@@ -126,29 +193,20 @@ def dev_send_1905(self, dest: str, message_type: int, *tlvs: Tuple[int, int, str
-
-        cmd = "DEV_SEND_1905,DestALid,{dest:s},MessageTypeValue,0x{message_type:04x}".format(**locals())
+        cmd = "DEV_SEND_1905,DestALid,{dest:s},MessageTypeValue,0x{message_type:04x}"\
+            .format(locals())
```
Why did you remove the unpacking? Now the format method receives a dict, and this code should fail...

arnout replied:
That was an accident, fixed.

tomereli commented on tests/test_flows.py:
```diff
@@ -234,6 +234,33 @@ def test_initial_ap_config(self):
+        # Regression test: MAC address should be case insensitive
+        mac_repeater1_upper = self.mac_repeater1.upper()
+        # Configure the controller and send renew
+        self.gateway_ucc.cmd_reply("DEV_RESET_DEFAULT")
```
Since autoconfiguration renew should be reused by multiple tests (4.4.3, 4.2.1, 4.8.1, etc.) - we should implement a helper function for ap_config_renew, preferable in a followup commit.
Something like:
```
def trigger_ap_config_renew(self, al_mac, credentials-2.4g, credentials-5g)
```

(nitpick) since this can be done in a separate PR, but I rather keep this comment open till its done.

tomereli commented on tests/test_flows.py:
```diff
@@ -261,6 +261,37 @@ def test_ap_config_renew(self):
+
+        time.sleep(3)
+        self.check_log(self.repeater1, "agent_wlan0", r".* tear down radio")
+        # TODO check that conn map was updated on controller
```
Why do we care about the connection map? The TODO should say that we should trigger topology query and check the AP operational BSS TLV which should be empty for all radios.

arnout replied:
There should be no need to trigger topology query, since the controller should do that automatically upon receiving a topology notification -> #925.

I added the TODO because I removed the corresponding code that existed in test_flows.sh, and then reimplemented it in a different way in the next commit. Since it wasn't a one-to-one mapping I thought it was better to do it in a separate commit, and I didn't want to bother with implementing it the original way first.

tomereli commented on tests/test_flows.py:
```diff
@@ -261,6 +261,37 @@ def test_ap_config_renew(self):
             self.fail("repeater1 didn't configure Multi-AP-24G-1")
 
+    def test_ap_config_bss_tear_down(self):
+        # Configure the controller and send renew
```
So this should already use a helper function.

tomereli commented on tests/test_flows.py:
```diff
@@ -133,6 +162,31 @@ def docker_command(self, device: str, *command: str) -> bytes:
         '''Execute `command` in docker container `device` and return its output.'''
         return subprocess.check_output(("docker", "exec", device) + command)
 
+    def beerocks_command(self, command: str) -> bytes:
```
`beerocks_cli_command` is more appropriate IMO (nitpick)

tomereli commented on tests/test_flows.py:
```diff
@@ -277,7 +331,16 @@ def test_ap_config_bss_tear_down(self):
         self.check_log(self.repeater1, "agent_wlan0", r"Received credentials for ssid: Multi-AP-24G-3 .* bss_type: 2")
         self.check_log(self.repeater1, "agent_wlan2", r".* tear down radio")
-        # TODO check that conn map was updated on controller
+        conn_map = self.get_conn_map()
```
So now I understand the conn map TODO comment, but still, we should imitate the exact steps which are tested in the testbed, so once #924 is available we should convert this to check the AP Operational BSS TLV instead of relying on the connection map.

arnout replied:
I disagree that we should do the exact steps of the testbed.

We should *at least* do the same as the testbed, but I think it's definitely OK if we do more.

tomereli commented on tests/test_flows.py:
```diff
@@ -245,7 +245,7 @@ def check_log(self, device: str, program: str, regex: str) -> bool:
         logfilename = os.path.join(self.rootdir, 'logs', device, 'beerocks_{}.log'.format(program))
         try:
-            # HACK check_log is often used immediately after triggering a message on the other side.
+            # HACK self.check_log(is, "often", rused immediately after triggering a message on the other side.)
```
What's up with this comment?

arnout replied:
Hahaha, search-replace effect!

tomereli commented on tests/test_flows.py:
```diff
@@ -453,6 +453,46 @@ def test_combined_infra_metrics(self):
+        sta_mac1 = '00:00:00:11:00:22'
+        sta_mac2 = '00:00:00:11:00:33'
+
+        self.gateway_ucc.dev_send_1905(self.mac_repeater1, 0x8009,
```
A comment explaining the 1905 command being sent will help understand the flow better.

tomereli reviewed (approved):
mainly nitpick comments, but overall I can't really fault this. LGTM

arnout committed:
tests: test_flows.py: make beerocks_cli_command more verbose

When sending a beerocks_cli command, print the command and the response
when in debug.

Note that this make the verbose mode really verbose because the conn map
is dumped a couple of times. But this actually can be useful for
debugging, so it's a good idea.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
